go-controller/pkg/ovn/egressfirewall_test.go

```diff
@@ -1014,13 +1108,122 @@ var _ = ginkgo.Describe("OVN EgressFirewall Operations", func() {
 
+					if !oldDNS {
+						// delete the dns name resolver object.
+						err = fakeOVN.fakeClient.OCPNetworkClient.NetworkV1alpha1().DNSNameResolvers(config.Kubernetes.OVNConfigNamespace).
+							Delete(context.TODO(), "dns-default", metav1.DeleteOptions{})
+						gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					}
+
 					// check dns address set is cleaned up on delete
 					gomega.Eventually(fakeOVN.nbClient).Should(libovsdbtest.HaveData(initialData))
 					return nil
 				}
 				err := app.Run([]string{app.Name})
 				gomega.Expect(err).NotTo(gomega.HaveOccurred())
-			})
+			},
+				table.Entry(fmt.Sprintf("correctly cleans up object that failed to be created using old dns resolution, gateway mode %s", gwMode), gwMode, true),
+				table.Entry(fmt.Sprintf("correctly cleans up object that failed to be created using new dns resolution, gateway mode %s", gwMode), gwMode, false),
+			)
+			table.DescribeTable("correctly creates egress firewall using different dns resolution methods and ip families", func(gwMode config.GatewayMode, oldDNS bool, resolvedIP string) {
+				if !oldDNS {
+					// enable the dns name resolver flag.
+					config.OVNKubernetesFeature.EnableDNSNameResolver = true
+				}
+				config.Gateway.Mode = gwMode
+				app.Action = func(ctx *cli.Context) error {
+					namespace1 := *newNamespace("namespace1")
+					dnsName := "a.b.c"
+					dnsNameLowerCaseFQDN := util.LowerCaseFQDN(dnsName)
+					egressFirewall := newEgressFirewallObject("default", namespace1.Name, []egressfirewallapi.EgressFirewallRule{
+						{
+							Type: "Allow",
+							To: egressfirewallapi.EgressFirewallDestination{
+								DNSName: dnsName,
+							},
+						},
+					})
+					startOvn(dbSetup, []v1.Namespace{namespace1}, nil)
+
+					var err error
+					if oldDNS {
+						setDNSOpsMock(dnsName, resolvedIP)
+						fakeOVN.controller.dnsNameResolver, err = dnsnameresolver.NewEgressDNS(fakeOVN.controller.addressSetFactory,
+							fakeOVN.controller.controllerName, fakeOVN.controller.stopChan, egressFirewallDNSDefaultDuration)
+						gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					} else {
+						// Initialize the dnsNameResolver.
+						fakeOVN.controller.dnsNameResolver, err = dnsnameresolver.NewExternalEgressDNS(fakeOVN.controller.addressSetFactory,
+							fakeOVN.controller.controllerName, true, fakeOVN.watcher.DNSNameResolverInformer().Informer())
+						gomega.Expect(err).NotTo(gomega.HaveOccurred())
+						err = fakeOVN.controller.dnsNameResolver.Run()
+						gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					}
+
+					// Create the egress firewall object.
+					_, err = fakeOVN.fakeClient.EgressFirewallClient.K8sV1().EgressFirewalls(egressFirewall.Namespace).
+						Create(context.TODO(), egressFirewall, metav1.CreateOptions{})
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+					if !oldDNS {
+						dnsNameResolver := newDNSNameResolverObject("dns-default", config.Kubernetes.OVNConfigNamespace, dnsNameLowerCaseFQDN,
```

Comment:
looks like this function could take resolvedIP as an argument instead of full status to simplify this call?